Sources/SwiftParser/Specifiers.swift

```diff
@@ -795,22 +796,56 @@ extension Parser {
       ) {
         let misspelledThrows = self.eat(handle)
         unexpected.append(misspelledThrows)
-        if effectSpecifiers?.throwsClause == nil {
-          if matchedSubset == S.CorrectThrowsTokenKinds.self {
-            synthesizedThrows = missingToken(spec)
-          } else {
-            synthesizedThrows = missingToken(.throws)
+        guard effectSpecifiers?.throwsClause == nil else {
+          continue
+        }
+        if matchedSubset == S.CorrectThrowsTokenKinds.self {
+          let synthesizedThrows = missingToken(spec)
+          var lookahead = self.lookahead()
+          guard lookahead.consume(if: .leftParen) != nil,
+            lookahead.canParseType()
+          else {
+            synthesizedThrowsClause = RawThrowsClauseSyntax(
+              throwsSpecifier: synthesizedThrows,
+              leftParen: nil,
+              type: nil,
+              rightParen: nil,
+              arena: self.arena
+            )
+            continue
           }
+
+          let leftParen = self.expect(.leftParen).token
```

Comment:
We shouldn’t be discarding the unexpected tokens from `expect`. I know that it can’t really happen here because we check in `lookahead.consume(if:)` that the next token is a left parenthesis but in general, dropping syntax nodes can lead to a syntax tree that doesn’t represent the original source code and it’s good practice to never drop anything. I’d change this to

```swift
let (unexpectedBeforeLeftParen, leftParen) = self.expect(.leftParen)
```
change `unexpected` to be of type `[RawSyntax]` (which requires a few conversions from `RawTokenSyntax` to `RawSyntax` using the `RawSyntax` initializer) and then below do

```swift
if let unexpectedBeforeLeftParen {
  unexpected.append(RawSyntax(unexpectedBeforeLeftParen))
}
unexpected.append(RawSyntax(leftParen))
unexpected.append(RawSyntax(type))
```

Notice how you also don’t need `allTokens` on `type` anymore.

And then you could do the same for `rightParen` to also use `expect`.